nntrainer/layers/multi_head_attention_layer.cpp

```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
+    weight_idx[AttentionParams::key_fc_bias] = context.requestWeight(
+      key_fc_bias_dim, weight_initializer, weight_regularizer,
+      weight_regularizer_constant, weight_decay, "key_fc_bias", true);
+  }
+
+  TensorDim value_fc_weight_dim(
+    {1, 1, value_width, num_heads * projected_value_dim_prop});
+  weight_idx[AttentionParams::value_fc_weight] = context.requestWeight(
+    value_fc_weight_dim, weight_initializer, weight_regularizer,
+    weight_regularizer_constant, weight_decay, "value_fc_weight", true);
+  if (!disable_bias) {
+    TensorDim value_fc_bias_dim(
+      {1, 1, 1, num_heads * projected_value_dim_prop});
+    weight_idx[AttentionParams::value_fc_bias] = context.requestWeight(
+      value_fc_bias_dim, weight_initializer, weight_regularizer,
+      weight_regularizer_constant, weight_decay, "value_fc_bias", true);
+  }
+
+  TensorDim fc_weight_dim(
+    {1, 1, num_heads * projected_value_dim_prop, output_shape});
+  weight_idx[AttentionParams::fc_weight] = context.requestWeight(
+    fc_weight_dim, weight_initializer, weight_regularizer,
+    weight_regularizer_constant, weight_decay, "fc_weight", true);
+  if (!disable_bias) {
+    TensorDim fc_bias_dim({1, 1, 1, output_shape});
+    weight_idx[AttentionParams::fc_bias] = context.requestWeight(
+      fc_bias_dim, weight_initializer, weight_regularizer,
+      weight_regularizer_constant, weight_decay, "fc_bias", true);
+  }
+
+  TensorDim projected_query_dim(
+    {batch_size, 1, query_height, num_heads * projected_query_dim_prop});
+  weight_idx[AttentionParams::projected_query] = context.requestTensor(
+    projected_query_dim, "projected_query", Tensor::Initializer::NONE, false,
+    TensorLifespan::FORWARD_FUNC_LIFESPAN);
+  weight_idx[AttentionParams::d_projected_query] = context.requestTensor(
+    projected_query_dim, "d_projected_query", Tensor::Initializer::NONE, false,
+    TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+  TensorDim projected_key_dim(
+    {batch_size, 1, key_height, num_heads * projected_key_dim_prop});
+  weight_idx[AttentionParams::projected_key] = context.requestTensor(
+    projected_key_dim, "projected_key", Tensor::Initializer::NONE, true,
+    TensorLifespan::FORWARD_FUNC_LIFESPAN);
+  weight_idx[AttentionParams::d_projected_key] = context.requestTensor(
+    projected_key_dim, "d_projected_key", Tensor::Initializer::NONE, false,
+    TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+  TensorDim projected_value_dim(
+    {batch_size, 1, value_height, num_heads * projected_value_dim_prop});
+  weight_idx[AttentionParams::projected_value] = context.requestTensor(
+    projected_value_dim, "projected_value", Tensor::Initializer::NONE, true,
+    TensorLifespan::FORWARD_FUNC_LIFESPAN);
+  weight_idx[AttentionParams::d_projected_value] = context.requestTensor(
+    projected_value_dim, "d_projected_value", Tensor::Initializer::NONE, false,
+    TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+
+  TensorDim transposed_projected_query_dim(
```

Comment:
Transpose matrix could be removed. please check it.

Reply:
For now NNTrainer does not support inplace transpose operation.
First thing to do is check transpose is needed and if it is we need to implement inplace transpose operation

Reply:
If the projected tensor can be overwritten, then it is better to use temporal tensor inside of transpose tensor op.

Reply:
Done